npeps/npep-133.md

```diff
@@ -0,0 +1,173 @@
+# NPEP-133: FQDN Selector for Egress Traffic
+
+* Issue:
+  [#133](https://github.com/kubernetes-sigs/network-policy-api/issues/133)
+* Status: Provisional
+
+## TLDR
+
+This enhancement proposes adding a new optional selector to specify egress peers
+using [Fully Qualified Domain
+Names](https://www.wikipedia.org/wiki/Fully_qualified_domain_name) (FQDNs).
+
+## Goals
+
+* Provide a selector to specify egress peers using a Fully Qualified Domain Name
+  (for example `kubernetes.io`).
+* Support basic wildcard matching capabilities when specifying FQDNs (for
+  example `*.cloud-provider.io`)
+* Currently only `ALLOW` type rules are proposed.
```

Comment:
Is all other egress denied, including in-cluster?

Reply:
Not exactly -- AdminNetworkPolicy doesn't operate on default deny semantics like that.

These rules would be paired with either a lower priority DENY rule, or use the default deny behavior of Kubernetes NetworkPolicy.

Reply:
@rahulkjoshi ^ is what I thought but the NPEP is not explicit in this regard. Might be worth adding a condensed form of ^ to be explicit.

Reply:
>These rules would be paired with either a lower priority DENY rule...

However, the NPEP states that `DENY` rules are not supported.

Reply:
Yeah I can see how that's confusing -- I hope long-term it's clearer when we provide the API but, I'll try to explain and I'm open to suggestions on how to phrase this so it's easier to understand for other readers:

**Quick Note, API details will be affected by what we decide in https://github.com/kubernetes-sigs/network-policy-api/pull/144**

Roughly speaking, the new API will extend the [AdminNetworkPolicyEgressPeer](https://github.com/kubernetes-sigs/network-policy-api/blob/774376b17a86ea0a4ebbd640aedab2ec000d2816/apis/v1alpha1/shared_types.go#L153-L178) by adding a new selector. AdminNetworkPolicyEgressPeer is specified inside the [AdminNetworkPolicyEgressRule](https://github.com/kubernetes-sigs/network-policy-api/blob/774376b17a86ea0a4ebbd640aedab2ec000d2816/apis/v1alpha1/adminnetworkpolicy_types.go#L157) struct (in the repeated field `To`).

When we say `DENY` rules are not supported, more specifically what we're saying is: If a `AdminNetworkPolicyEgressRule` has `Action: "Deny"`, then none of the `AdminNetworkPolicyEgressPeer`s in that rule can specify FQDN Selectors.

You can however, specify another `AdminNetworkPolicyEgressRule` with `Action: "Deny"` (you can either make another entry in `AdminNetworkPolicySpec.Egress` or create a separate ANP object) and include other selectors in that rule.